### PR TITLE
[yamlcomposer] Add automatic regeneration when referenced environment variables changed

### PR DESCRIPTION
**Pull request by jimtng**
# Overview

This pull request introduces automatic detection and recompilation of YAML configuration files when any referenced environment variables change between openHAB sessions.

## Summary of Changes

1. **Environment Tracking**: Expands tracking to capture all environment variables referenced via `${ENV.KEY}` or `${ENV['KEY']}` syntax, including those within included files, inserted templates, and merge-key substitutions.

1. **Header Metadata**: Injects `Env-Deps` and `Env-Hash` fields into the generated file headers to record dependencies and their cryptographic values.

1. **Change Detection & Regeneration**: Adds ComposerUtils#isEnvironmentChanged to compare current environment values against the stored hash in the header, triggering a clean automatic regeneration when values drift.

### CODE REVIEW

File: bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java [hunk 1]
```diff
@@ -93,13 +121,170 @@ static void writeCompiledOutput(Object dataMap, Path sourcePath, Path outputPath
...
+                }
+
+                // Stop early if we have the hash and aren't collecting dependencies anymore
+                if (!envHashLine.isEmpty() && !collectingEnvDeps) {
+                    break;
+                }
+            }
+
+            // If header markers are missing (legacy file), trigger regeneration
+            if (envHashLine.isEmpty()) {
```
Comment (wborn) [inline bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java]: Newly generated files without any environment dependencies also have an empty `Env-Hash`, because `computeEnvHash()` returns an empty string for an empty dependency list. This check therefore treats those files as legacy on every startup and forces them to be regenerated even when neither the source nor any includes changed.

Could we distinguish between the `Env-Hash` header being absent and it being present with an empty value? Alternatively, `computeEnvHash()` could return a real hash for the empty dependency set.

It would also be good to add a test that writes an output with no tracked environment variables and verifies that `isEnvironmentChanged()` returns `false`.

File: bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java [hunk 2]
```diff
@@ -93,13 +121,170 @@ static void writeCompiledOutput(Object dataMap, Path sourcePath, Path outputPath
...
+    }
+
+    private static String computeEnvHash(List<String> envVars, Map<String, String> envMap) {
+        if (envVars.isEmpty()) {
+            return "";
+        }
+        try {
+            MessageDigest digest = MessageDigest.getInstance("SHA-256");
+            for (String varName : envVars) {
+                String val = envMap.get(varName);
```
Comment (wborn) [inline bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java:280]: The serialization used as input to SHA-256 is ambiguous and also does not distinguish an unset environment variable from one explicitly set to an empty string.

For example, both an unset value and `""` append no value bytes, so changing `FOO` from unset to empty will produce the same hash. For a direct `${ENV.FOO}` substitution those values can be semantically different.

The `=` / `;` delimiters can also produce the same byte stream for different value combinations because environment values may contain those characters. For variables `A` and `B`, these two sets produce the same serialized input:

- `A=""`, `B="x;B=y"`
- `A=";B=x"`, `B="y"`

Could we use an unambiguous representation, such as length-prefixed names/values together with an explicit presence marker? A test for the unset ↔ empty case would cover the most important edge case.

**Review by wborn — DISMISSED** [review]
AI review performed before manual maintainer review.

The overall approach looks sound, and the environment dependency tracking is propagated consistently through substitutions, conditionals, includes, templates, and merge-key processing. However, AI found two correctness issues in the environment change detection that can either cause unnecessary regeneration or miss a required regeneration. Please see the inline comments.

CI is currently green. Human maintainer review is still required.

File: bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java (hunk 1 above)
Comment (jimtng) [reply bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java]: > Could we distinguish between the `Env-Hash` header being absent and it being present with an empty value?

Implemented this.



> It would also be good to add a test that writes an output with no tracked environment variables and verifies that `isEnvironmentChanged()` returns `false`

done

File: bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java (hunk 2 above)
Comment (jimtng) [reply bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java:280]: > Could we use an unambiguous representation, such as length-prefixed names/values together with an explicit presence marker? A test for the unset ↔ empty case would cover the most important edge case.

done

**Comment by jimtng** [comment]
@HolgerHees I’ve added a doc section describing the differences between Composer’s ENV and Core ENV. Could you take a look and let me know if anything should be improved?

https://github.com/jimtng/openhab-addons/blob/57135c9fadc321682ed7643113b4af225653d2d9/bundles/org.openhab.io.yamlcomposer/doc/variables.md#env-to-access-environment-variables

**Comment by HolgerHees** [comment]
> @HolgerHees I’ve added a doc section describing the differences between Composer’s ENV and Core ENV. Could you take a look and let me know if anything should be improved?

looks good to me.

**Review by wborn — COMMENTED** [review]
The two correctness issues from the previous AI review have been addressed in the current revision.

The generated header now correctly distinguishes files with no environment dependencies from legacy files without the environment metadata, so unchanged files are no longer regenerated unnecessarily. The environment hash serialization also now uses explicit presence markers and length-prefixed names and values, correctly distinguishing unset variables from empty values and avoiding ambiguous hash inputs.

The environment dependency tracking is propagated consistently through substitutions, conditionals, includes, templates, and merge-key processing. No further issues were found in the current changes.

CI is green. Human maintainer review is still required.

**Comment by lsiepel** [comment]
When the conflicts are resolved i'll have a final look before we merge. I have no comments atm

**Comment by jimtng** [comment]
rebased

**Comment by jimtng** [comment]
@lsiepel Following up on your offer to review — whenever you have time, this one will help avoid some rebase noise later. Thanks!

File: bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/expression/ExpressionEvaluator.java [hunk 3]
```diff
@@ -141,20 +146,49 @@ public static boolean isTruthy(@Nullable Object value) {
...
+
+        @Override
+        public @Nullable String get(@Nullable Object key) {
+            if (key instanceof String varName) {
+                callback.accept(varName);
+                return System.getenv(varName);
+            }
+            return null;
+        }
+
```
Comment (lsiepel) [inline bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/expression/ExpressionEvaluator.java:199]: `${ENV...}` is exposed as a `Map`, and the documentation permits calling Java methods on variable values. `TrackingEnvMap` only records `get()`: inherited `AbstractMap.containsKey()` scans `entrySet()` without invoking the callback. Consequently `${ENV.containsKey('FEATURE_FLAG')}` can change after a restart while `Env-Deps` remains empty, so the existing compiled YAML is considered current and stays stale. Please track direct presence checks as well (and cover this access form with a test).

```suggestion
        @Override
        public boolean containsKey(@Nullable Object key) {
            if (key instanceof String varName) {
                callback.accept(varName);
            }
            return System.getenv().containsKey(key);
        }
```

File: bundles/org.openhab.io.yamlcomposer/src/test/java/org/openhab/io/yamlcomposer/internal/YamlComposerEnvironmentVariableTest.java [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+            try (MockedStatic<OpenHAB> openHABMock = mockOpenHabMetadata()) {
+                ComposerUtils.writeCompiledOutput(yamlObject, main, output, trackedEnv);
+            }
+
+            assertThat(Files.exists(output), is(true));
+            assertThat(Files.readString(output), containsString("MY_TEST_ENV"));
+        }
+
+        @Test
+        @DisplayName("Returns false initially and true when a tracked environment variable value changes")
```
Comment (lsiepel) [inline bundles/org.openhab.io.yamlcomposer/src/test/java/org/openhab/io/yamlcomposer/internal/YamlComposerEnvironmentVariableTest.java:222]: The two earlier implementation fixes are present, but their requested regression cases are not: this test only covers one non-empty value changing to another. Please also verify that an output with zero tracked variables is not treated as legacy, and that changing one tracked name from absent to an empty value is detected. Those assertions would prevent both corrected branches from regressing.

File: bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java [hunk 5]
```diff
@@ -93,13 +122,178 @@ static void writeCompiledOutput(Object dataMap, Path sourcePath, Path outputPath
...
+
+        try (BufferedReader reader = Files.newBufferedReader(outputPath, StandardCharsets.UTF_8)) {
+            String line;
+            List<String> envVars = new ArrayList<>();
+            String envHashLine = null;
+            boolean collectingEnvDeps = false;
+            boolean headerStarted = false;
+
+            while ((line = reader.readLine()) != null) {
+                // Check for separator lines to determine if we're in the header section
```
Comment (lsiepel) [inline bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java]: This method's new comments mostly narrate the following condition or return (`second separator`, `stop early`, `missing markers`, `no variables`, `safe fallback`). They obscure the parser more than they explain it. Please retain only comments that document a non-obvious header-format invariant and let the state names/control flow explain the rest.

**Review by lsiepel — COMMENTED** [review]
See the inline comments for details. 

_This is an AI-assisted review._

File: bundles/org.openhab.io.yamlcomposer/src/test/java/org/openhab/io/yamlcomposer/internal/YamlComposerEnvironmentVariableTest.java (hunk 4 above)
Comment (jimtng) [reply bundles/org.openhab.io.yamlcomposer/src/test/java/org/openhab/io/yamlcomposer/internal/YamlComposerEnvironmentVariableTest.java:222]: tests added

File: bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java (hunk 5 above)
Comment (jimtng) [reply bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java]: comments removed except about legacy file

File: bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/expression/ExpressionEvaluator.java (hunk 3 above)
Comment (jimtng) [reply bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/expression/ExpressionEvaluator.java:199]: Good catch. I couldn't come up with an ideal solution without a convoluted scheme to address the following edge case:  When the user wants the value for an actual env var named `containsKey`, it won't be tracked. 

I feel that this is a rare enough edge case not to bother with a more complex solution to address it.
Currently this is not mentioned in the user-facing documentation for the same reason that it is probably a rare edge case. It is documented in a source code comment

File: bundles/org.openhab.io.yamlcomposer/src/test/java/org/openhab/io/yamlcomposer/internal/YamlComposerEnvironmentVariableTest.java (hunk 4 above)
Comment (lsiepel) [reply bundles/org.openhab.io.yamlcomposer/src/test/java/org/openhab/io/yamlcomposer/internal/YamlComposerEnvironmentVariableTest.java:222]: The zero-dependency case is covered now, but the requested absent-to-empty transition is still missing. returnsFalseInitiallyAndTrueWhenEnvironmentValueChanged() only covers one non-empty value changing to another. Please add a regression assertion that writes the header while the name is absent, adds the same name with "", and verifies that isEnvironmentChanged() returns true.

File: bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/expression/ExpressionEvaluator.java [hunk 6]
```diff
@@ -141,20 +146,66 @@ public static boolean isTruthy(@Nullable Object value) {
...
+        @Override
+        public @Nullable String get(@Nullable Object key) {
+            if (key instanceof String varName) {
+                // Special case: Ignore "containsKey" during EL property probing so EL falls back
+                // to calling TrackingEnvMap.containsKey("VAR") instead of tracking "containsKey".
+                // Caveat: Fetching the value of an actual env var named "containsKey" (${ENV.containsKey})
+                // will return the value, but it won't be tracked.
+                // (Existence checks like `'containsKey' in ENV` still work).
+                if ("containsKey".equals(varName)) {
```
Comment (lsiepel) [inline bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/expression/ExpressionEvaluator.java]: The `containsKey` special case does not cover other supported `Map` methods. For `${ENV.size()}`, Jinjava calls `get("size")`, so the current implementation records `size` as the dependency and then evaluates the actual environment size through `entrySet()`. Adding or removing another environment variable therefore changes the generated value without changing `Env-Hash`, leaving stale output. `keySet()`, `values()`, `entrySet()`, and `containsValue()` have the same problem. Please represent whole-environment access explicitly and add a regression test.

File: bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java [hunk 7]
```diff
@@ -93,13 +122,175 @@ static void writeCompiledOutput(Object dataMap, Path sourcePath, Path outputPath
...
+        } catch (IOException e) {
+            return true;
+        }
+    }
+
+    private static List<String> parseEnvVars(String content) {
+        if (content.isEmpty()) {
+            return List.of();
+        }
+        return Arrays.stream(content.split(",")).map(String::trim).filter(s -> !s.isEmpty()).toList();
```
Comment (lsiepel) [inline bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java]: `Env-Deps` is not round-trippable because names are comma-delimited without escaping. Environment-variable names containing commas are valid on [POSIX](https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/V1_chap08.html) and [Windows](https://learn.microsoft.com/en-us/windows/win32/procthread/environment-variables), and bracket syntax makes them addressable. For `ENV['YAML_COMPOSER,REVIEW']`, the writer hashes one name but `parseEnvVars()` reads two names. The current implementation consequently reports a change against the identical environment and regenerates on every startup. Please use an unambiguous encoding and cover this round trip.

**Review by lsiepel — CHANGES_REQUESTED** [review]
One of the previous comments was not fully fixed. Also found these two.

File: bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java (hunk 7 above)
Comment (jimtng) [reply bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java]: good catch, I didn't know commas are valid env var names. I've adopted urlencode/decode to deal with it.

File: bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/expression/ExpressionEvaluator.java (hunk 6 above)
Comment (jimtng) [reply bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/expression/ExpressionEvaluator.java]: I have clarified the ENV tracking scope in the documentation, and updated the implementation / test accordingly.

Specifically:
> If a Composer source file references environment variables via standard lookups (e.g., `${ENV.VAR_NAME}` or `${ENV['VAR_NAME']}`), **changes to those variable values will automatically trigger regeneration** of the compiled YAML during openHAB startup.
>
> Advanced operations on `ENV`—such as checking key existence (`'VAR' in ENV` or `ENV.containsKey(...)`), iterating over the map, or querying map properties (e.g., `ENV.size()`)—are not tracked for auto-regeneration.

If we were to support tracking of ENV.size and various other advanced methods, it would be insufficient to merely track the env names + their hashes. We would need to also keep track of the _entire_ ENV map contents in order to verify changes in size, values, etc. This is not practical, if not difficult/complex, simply to cater to rare edge cases. Most env usage in YAML config simply needs to access particular keys which was the original intention of this PR.

File: bundles/org.openhab.io.yamlcomposer/src/test/java/org/openhab/io/yamlcomposer/internal/YamlComposerEnvironmentVariableTest.java (hunk 4 above)
Comment (jimtng) [reply bundles/org.openhab.io.yamlcomposer/src/test/java/org/openhab/io/yamlcomposer/internal/YamlComposerEnvironmentVariableTest.java:222]: I've added an absent-to-empty test case

File: bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java (hunk 7 above)
Comment (lsiepel) [reply bundles/org.openhab.io.yamlcomposer/src/main/java/org/openhab/io/yamlcomposer/internal/ComposerUtils.java]: Credits to codex. ;-)

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM
